chaos_genius/views/download_view.py

```diff
@@ -0,0 +1,215 @@
+            raise Exception(f"No anomaly data found for KPI id {kpi_id}")
+
+        def row_gen(data_points):
+            csv_headers = [
+                "datetime",
+                "value",
+                "severity",
+                "upper_bound",
+                "lower_bound",
+                "series_type",
+                "series_name",
+            ]
+            yield csv_headers
+            for row in data_points:
+                attr_list = [
+                    row.data_datetime.strftime("%a %-d %B %H:%M:%S %Y"),
+                    str(row.y),
+                    str(row.severity),
+                    str(row.yhat_upper),
+                    str(row.yhat_lower),
+                    str(row.anomaly_type),
+                    str(row.series_type),
+                ]
+                yield attr_list
+
+        response = Response(iter_csv(row_gen(data_points)), mimetype="text/csv")
+        response.headers[
+            "Content-Disposition"
+        ] = f"attachment; filename=KPI-{kpi_id}-anomaly-data.csv"
+        return response
+    except Exception as e:
+        return jsonify(
+            {"status": "failure", "message": f"Downloading data failed: {e}"}
+        )
+
+
+@blueprint.route("/<int:kpi>/chart_data", methods=["GET"])
+def kpi_download_line_data(kpi: int):
+    """API endpoint to download chart data for last 60 days."""
+    try:
+        line_data = kpi_line_data(kpi, download=True)
+        if line_data[0] == "error":
+            raise Exception(f"{line_data[1]}")
+        data_points = line_data[2]
+
+        if not data_points:
+            raise Exception(f"No chart data found for KPI id {kpi}")
+
+        def row_gen(data_points):
+            csv_headers = ["date", "value"]
+            yield csv_headers
+            for row in data_points:
+                attr_list = [row["date"].strftime("%a %-d %B %Y"), str(row["value"])]
+                yield attr_list
+
+        response = Response(iter_csv(row_gen(data_points)), mimetype="text/csv")
+        response.headers[
+            "Content-Disposition"
+        ] = f"attachment; filename=KPI-{kpi}-panel-chart-data.csv"
+        return response
+    except Exception as e:
+        return jsonify(
+            {"status": "failure", "message": f"chart data download failed: {e}"}
+        )
+
+
+@blueprint.route("/<int:kpi_id>/hierarchical_data", methods=["GET"])
+def download_hierarchical_data(kpi_id):
+    """API endpoint to download RCA hierarchical data."""
+    try:
+        timeline = request.args.get("timeline")
+        if not timeline:
+            raise Exception("Please provide timeline as an argument")
+
+        kpi_info = get_kpi_data_from_id(kpi_id)
+        dimensions = kpi_info.get("dimensions")
+        if not dimensions:
+            raise Exception(f"Failed to fetch dimensions for kpi {kpi_id}")
+
+        data_list = {}
+        for dimension in dimensions:
+            result = rca_hierarchical_data(kpi_id, timeline, dimension)
```

Comment:
Here, a new query is made for each dimension (inside `rca_hierarchical_data`). It would be better to get data for all dimensions and then group it according to dimensions after retrieving.

Alternatively, you can add context to the generator using [this](https://flask.palletsprojects.com/en/2.1.x/patterns/streaming/#streaming-with-context) and do the query inside the generator (`row_gen`). Although, this will require adding error handling inside the generator too.

What do you think?